src/components/Form.tsx

```diff
@@ -4,63 +4,96 @@ import Button from "./Button";
+  warningText?: string | JSX.Element | Array<JSX.Element>;
 }
 
 export default class Form extends React.Component<FormProps, {}> {
+  private messageRef = React.createRef<HTMLParagraphElement>();
+  private formRef = React.createRef<HTMLFormElement>();
+
   constructor(props: FormProps) {
     super(props);
     this.submit = this.submit.bind(this);
     this.message = this.message.bind(this);
   }
 
-  componentDidUpdate() {
-    if (this.refs["errorMessage"]) {
-      (this.refs["errorMessage"] as HTMLElement).focus();
+  componentDidUpdate(oldProps: FormProps) {
+    if (this.messageRef.current && this.messageRef.current.focus) {
+      // If the message hasn't changed (e.g. the user has typed in new input
+      // but not re-submitted the form yet), there's no point re-focusing on it.
+      let messageProps = ["errorText", "successText", "warningText"];
+      let hasNewMessage = messageProps.some(prop => oldProps[prop] !== this.props[prop]);
+      hasNewMessage && this.messageRef.current.focus();
     }
   }
 
-  submit(event: React.MouseEvent): void {
+  async submit(event: React.MouseEvent<HTMLButtonElement>) {
     event.preventDefault();
-    let form = (this.refs["form"] as any);
+    let form = this.formRef.current;
     const data = new (window as any).FormData(form);
-    this.props.onSubmit(data);
+    await this.props.onSubmit(data);
   };
 
-  message(text: string, type: string): JSX.Element {
-    return (
-      <p className={`alert alert-${type}`} role="alert" ref={`${type}Message`} tabIndex={-1}>
-        {text}
-      </p>
+  message(text, messageType: string) {
+    let messageProps = {
+      className: `alert alert-${messageType}`,
+      role: "alert",
+      ref: this.messageRef,
+      tabIndex: -1
+    };
+
+    return React.createElement(
+      text.type || "p",
+      {...messageProps, ...text.props},
+      text
     );
   }
 
   render(): JSX.Element {
-    const formClass = `clearfix${this.props.className ? " " + this.props.className : ""}`;
-
+    let className = "clearfix" + (this.props.className ? ` ${this.props.className}` : "");
     return(
-      <form ref="form" className={formClass}>
+      <form ref={this.formRef} className={className}>
         {
           this.props.errorText && this.message(this.props.errorText, "danger")
         }
         {
-          this.props.successText && !this.props.errorText &&
-          this.message(this.props.successText, "success")
+          this.props.successText && !this.props.errorText && this.message(this.props.successText, "success")
         }
         {
           this.props.infoText && this.message(this.props.infoText, "info")
         }
+        {
+          this.props.loadingText && this.message(this.props.loadingText, "loading")
+        }
+        {
+          this.props.warningText && this.message(this.props.warningText, "warning")
+        }
         { this.props.title &&
           <label className="form-title">{this.props.title}</label>
```

Comment:
The more I look at this the more I think a `label` isn't right. Specifically, after looking at the change for the Change Admin Password page, it seems like it _could_ be a heading or a fieldset and this be a `legend` instead. The previous version of the Change Admin Password had an h2 and the legend in the fieldset in the form described the form. Now the label describes the form but there is nothing connecting this label with the form, and the h2 is lost. I suggest creating a ticket to talk about the form structure with Willa to keep this PR solely for replacing the component. But overall I'd like her to review how well this updated structure is.

Reply:
Change Admin Password in `circulation-web`.

Reply:
We can't have the `Form` component automatically turn the `title` prop into a legend, because that assumes that, in every single implementation, the developer has passed exactly one fieldset to the `content` prop, and that's clearly not realistic.  I agree that it should be a legend for the Change Admin Password form (and anything else that used to have an h2), though...and that would follow the uswds patterns.  I'll change those back in the circ-web repo.  As for this repo, I think it makes more sense to wait and have the form structure conversation (and involve Willa) when we're working on updating/importing the reusable `Fieldset` component.

Reply:
Right, but it could be written such that if there is only one fieldset that is passed (or none), this would be converted to a fieldset and the title a legend. Otherwise just use as many fieldsets as were passed. I don’t think the `fieldset` element needs to be its own component. It doesn’t seem to add anything more than styling which is already covered by the style rules. Unless if there’s a big need and reason to create a fieldset component. There are other components that should be worked on instead.

Reply:
The fieldset doesn't add more than styling; I just figured that (in the spirit of atomic design!), having a reusable fieldset might encourage people to use fieldsets!   But more importantly: I'm concerned that, if we automatically convert everything to a fieldset, we're just going to end up with fieldsets containing stuff that shouldn't actually be grouped together (e.g. I could see how you would want a form with individual input fields that wouldn't make sense to be grouped together, but you wouldn't have wrapped each one in its own fieldset), and that will be confusing to users, especially from an accessibility standpoint.  Anyway.  A conversation for another, more fieldset-focused, day!

Reply:
Yes, but we should not be building forms using fieldsets and grouping things that don’t belong together in the first place. Fieldsets are basic HTML elements and don’t think it’s the same for atomic design. I just don’t see why it’s necessary or useful but we’ll see. I just don’t think having something like 
```
<Form
  content={
    <Fieldset  content={…} >
       ….
    </Fieldset>
>
</Form>
```
saves us anything instead of doing 
```
<form>
   <fieldset>
     …
   </fieldest>
</form>
```